### PR TITLE
Add support for 'U' uppercase

### PR DESCRIPTION
**Pull request by rebornix**
Convert selected range to Uppercase is an often-used command.

The code is well designed so it's easy enough to add new commands support. The only catch is all commands are implemented inside a single file, which makes it lack of categorization and readability, any ideas to refactor it?

### CODE REVIEW

**Comment by johnfn** [comment]
Hey @rebornix, thanks for the awesome PR! I'm glad to hear that the code is easy to understand for the most part.

I hear you about the gigantic file. Honestly, for whatever reason, I've always had a lower sensitivity to this issue than other people, which is why I haven't split it yet. When I tried briefly, I was having difficulty with decorators and splitting actions.ts into multiple files - I think TypeScript decorators may have some implementation bugs (they are classified as experimental, after all). 

TL;DR - I'm totally open to PRs that split it up. It's a task I plan to do at some point, but it's not high priority. If a couple people nag me to do it, I'll bump the priority 😉

File: src/actions/actions.ts [hunk 1]
```diff
@@ -43,6 +43,42 @@ const compareKeypressSequence = function (one: string[], two: string[]): boolean
   return true;
 };
 
+const adjustVimSelectionPosition = function(vimState: VimState, start: Position, end: Position): any {
```
Comment (johnfn) [inline src/actions/actions.ts]: Two comments here:

* `adjustVimSelectionPosition` is an ambiguous name. Something like `transformRangeForDeleteOperator` might be better. 
* Instead of `any` I would prefer using a TypeScript tuple type: `[Position, Position]`

Comment (johnfn) [inline src/actions/actions.ts]: Also I generally prefer static functions to free-floating ones. When we split this file up into many files, the decision will make our lives easier. 😃

File: src/actions/actions.ts [hunk 2]
```diff
@@ -583,6 +589,21 @@ export class DeleteOperatorXVisual extends BaseOperator {
 }
 
 @RegisterAction
+export class UpperCaseOperatorXVisual extends BaseOperator {
+    public keys = ["U"];
+    public modes = [ModeName.Visual, ModeName.VisualLine];
+    
+    public async run(vimState: VimState, start: Position, end: Position): Promise<VimState> {
+      [start, end] = adjustVimSelectionPosition(vimState, start, end);
```
Comment (johnfn) [inline src/actions/actions.ts]: I actually don't understand why you need to transform the range here. Most of the logic in DeleteOperator is to ensure we get exactly the right newlines we need to delete, but we don't really care about capitalizing newlines. 

I think just running from start to end.getRight() should be enough?

File: src/actions/actions.ts [hunk 3]
```diff
@@ -583,6 +589,21 @@ export class DeleteOperatorXVisual extends BaseOperator {
 }
 
 @RegisterAction
+export class UpperCaseOperatorXVisual extends BaseOperator {
```
Comment (johnfn) [inline src/actions/actions.ts]: You can just name this UpperCaseOperator.

**Comment by johnfn** [comment]
Last thing - would you mind adding a few tests?

File: src/actions/actions.ts (hunk 1 above)
Comment (rebornix) [inline src/actions/actions.ts]: Thanks for the naming suggestion, it's always pain for non-English speaker like me :) The strict return type suggestion is also good, will change it later.

File: src/actions/actions.ts (hunk 2 above)
Comment (rebornix) [inline src/actions/actions.ts]: When users select ranges in reverse order, we need to swap `start` and `end` first, this is contained in the extracted method so I just reuse it. 

Moreover, this method can be used for any Visual/VisualLine selection and I think it would be perfect to deal every special case if possible, then people don't need to worry about when to `getRight` or `getLeft`, whether to take current character under cursor into account or not, etc.

**Comment by rebornix** [comment]
@johnfn thanks for your detailed review and suggestion. I tried to run test cases on my Windows Machine previously but seems VS Code package will download Linux related package instead. I'll look into it and add tests :smiley:

File: src/actions/actions.ts (hunk 2 above)
Comment (johnfn) [inline src/actions/actions.ts]: Ah; all that logic is contained within executeOperator inside modeHandler.ts.

However, I think that only the swapping of start and stop should be moved in there (that's a good catch, btw!) - the rest of this stuff is DeleteOperator specific.

Comment (rebornix) [inline src/actions/actions.ts]: Cool, let me see if I can add this refactor together with this pr.

Comment (johnfn) [inline src/actions/actions.ts]: 👍  I don't think that you'll end up needing to refactor out this function at all.

Comment (johnfn) [inline src/actions/actions.ts]: (Though if you want you can remove the start/stop switching logic I have in the other operators...)

Comment (rebornix) [inline src/actions/actions.ts]: Will do that. Once the code is updated you can know if I fully get your ideas :)

BTW, send you an email about other funny stuff.

**Comment by rebornix** [comment]
I've tried to move the start/end swapping for Visual mode but failed. The reason is that this will breaks deletion.

1. Delete Operator may be triggered by `executeOperator` or `runAction`, while we don't have any position information inside `runAction`, so we have to keep the swapping inside Delete Operator.
2. With above info in mind, we have the swapping inside Delete Op. So putting the swapping inside `executeOperator` will lead to the duplication of swapping for Delete Op as `executeOperator` might trigger Delete Op as well.

So the temporary solution for UpperCase operator is reusing the swapping method. This is caused by my limited knowledge of Action/Operator relationships and the workflow, once I have better understanding of the code, I may come up with a prettier solution.

File: src/actions/actions.ts [hunk 4]
```diff
@@ -229,6 +229,17 @@ export class BaseOperator extends BaseAction {
      * Run this operator on a range, returning the new location of the cursor.
      */
     run(vimState: VimState, start: Position, stop: Position): Promise<VimState> { return; }
+
+    public transformRange(start: Position, end: Position): [Position, Position] {
```
Comment (rebornix) [inline src/actions/actions.ts]: This needs to be reconsidered whether to put the method here.

**Comment by johnfn** [comment]
Hmm, I see. I will dig into this tomorrow and see what I can figure out!
